File: studio/repositories/templates_repo.py
"""Generated personalized template persistence."""

import time
import uuid

from studio import db
from studio.models import GenTemplate


def _public(row):
    from studio import draft

    if not row:
        return None
    if isinstance(row, dict):
        clean = dict(row)
        clean.pop("_id", None)
        if clean.get("htmlContent"):
            clean["htmlContent"] = draft.sanitize_html(clean["htmlContent"])
        return clean
    clean = {
        **(row.doc or {}),
        "templateId": row.templateId,
        "projectId": row.projectId,
        "galleryIndex": row.galleryIndex,
        "generatedAt": row.generatedAt,
    }
    if clean.get("htmlContent"):
        clean["htmlContent"] = draft.sanitize_html(clean["htmlContent"])
    return clean


def replace_for_project(project_id, templates):
    now = time.time()
    saved = []
    with db.session() as s:
        s.query(GenTemplate).filter_by(projectId=project_id).delete()
        for idx, item in enumerate(templates):
            doc = dict(item)
            template_id = doc.get("templateId") or uuid.uuid4().hex[:12]
            row = GenTemplate(
                templateId=template_id,
                projectId=project_id,
                galleryIndex=doc.get("galleryIndex", idx),
                generatedAt=now,
                doc=doc,
            )
            s.add(row)
            saved.append(_public(row))
        s.commit()
    saved.sort(key=lambda d: (d.get("galleryIndex", 999), d.get("templateId", "")))
    return saved


def list_for_project(project_id):
    with db.session() as s:
        rows = (
            s.query(GenTemplate)
            .filter_by(projectId=project_id)
            .order_by(GenTemplate.galleryIndex, GenTemplate.templateId)
            .all()
        )
        return [_public(r) for r in rows]


def get(project_id, template_id):
    with db.session() as s:
        row = (
            s.query(GenTemplate)
            .filter_by(projectId=project_id, templateId=template_id)
            .first()
        )
        return _public(row)

File: src/server/repos/templatesRepo.ts
import { Prisma } from "@/generated/prisma/client";
import { sanitizeHtml } from "../draft";
import { prisma } from "../db";
import { toDoc } from "./doc";

function publicTemplate(row: {
  templateId: string;
  projectId: string | null;
  galleryIndex: number | null;
  generatedAt: number | null;
  doc: unknown;
}): Record<string, unknown> {
  const clean: Record<string, unknown> = {
    ...((row.doc as Record<string, unknown> | null) ?? {}),
    templateId: row.templateId,
    projectId: row.projectId,
    galleryIndex: row.galleryIndex,
    generatedAt: row.generatedAt,
  };
  if (typeof clean.htmlContent === "string") {
    clean.htmlContent = sanitizeHtml(clean.htmlContent);
  }
  return clean;
}

export async function replaceForProject(
  projectId: string,
  templates: Record<string, unknown>[],
): Promise<Record<string, unknown>[]> {
  const now = Date.now() / 1000;
  await prisma().builderTemplate.deleteMany({ where: { projectId } });
  const saved: Record<string, unknown>[] = [];
  for (let idx = 0; idx < templates.length; idx++) {
    const item = { ...templates[idx]! };
    const templateId = (item.templateId as string | undefined) ?? randomId();
    const row = await prisma().builderTemplate.create({
      data: {
        templateId,
        projectId,
        galleryIndex: (item.galleryIndex as number | undefined) ?? idx,
        generatedAt: now,
        doc: item as Prisma.InputJsonValue,
      },
    });
    saved.push(publicTemplate(row));
  }
  saved.sort(
    (a, b) =>
      ((a.galleryIndex as number | undefined) ?? 999) -
        ((b.galleryIndex as number | undefined) ?? 999) ||
      String(a.templateId).localeCompare(String(b.templateId)),
  );
  return saved;
}

export async function listForProject(projectId: string): Promise<Record<string, unknown>[]> {
  const rows = await prisma().builderTemplate.findMany({
    where: { projectId },
    orderBy: [{ galleryIndex: "asc" }, { templateId: "asc" }],
  });
  return rows.map((row) => publicTemplate(row));
}

export async function get(
  projectId: string,
  templateId: string,
): Promise<Record<string, unknown> | null> {
  const row = await prisma().builderTemplate.findFirst({ where: { projectId, templateId } });
  return row ? publicTemplate(row) : null;
}

function randomId(): string {
  const buf = new Uint8Array(6);
  crypto.getRandomValues(buf);
  return Array.from(buf, (b) => b.toString(16).padStart(2, "0")).join("");
}
